### PR TITLE
unit rvalue, use constant `()` instead of tuple

### LINKED ISSUE #70886: We should use a () constant instead of `Rvalue::Aggregate(box AggregateKind::Tuple, vec![])`.
This should be changed to return a `()` `Operand::Const` (*not* `Rvalue::Use` of that operand): https://github.com/rust-lang/rust/blob/39b62533c7f9d0581a6ea9b9fc2cc51f21c3b5b0/src/librustc_mir_build/build/misc.rs#L35-L37

These two other places should switch to a `()` `Rvalue::Use(Operand::Const(...))`: https://github.com/rust-lang/rust/blob/39b62533c7f9d0581a6ea9b9fc2cc51f21c3b5b0/src/librustc_mir_build/build/cfg.rs#L66 https://github.com/rust-lang/rust/blob/39b62533c7f9d0581a6ea9b9fc2cc51f21c3b5b0/src/librustc_mir/transform/promote_consts.rs#L838

This issue has been assigned to @lcnr via [this comment](https://github.com/rust-lang/rust/issues/70886#issuecomment-610439401).

### PR DESCRIPTION
**Pull request by lcnr**
fixes #70886

r? @eddyb

### CODE REVIEW

**Comment by eddyb** [comment]
LGTM, but let's wait for #70721 first (since `mir-opt` tests are affected).

File: src/librustc_middle/mir/mod.rs [hunk 1]
```diff
@@ -2004,6 +2004,15 @@ impl<'tcx> Debug for Operand<'tcx> {
 }
 
 impl<'tcx> Operand<'tcx> {
+    /// Creates a `()` operand.
+    pub fn unit(tcx: TyCtxt<'tcx>, span: Span) -> Self {
```
Comment (Centril) [inline src/librustc_middle/mir/mod.rs:2008]: I'd appreciate avoiding new inherent methods on MIR data types that refer to the TyCtxt.

Comment (lcnr) [reply src/librustc_middle/mir/mod.rs:2008]: Hm, I guess we could add a unit constant as a field to the building context instead :thinking: 

cc @eddyb

Comment (eddyb) [reply src/librustc_middle/mir/mod.rs:2008]: Or keep the old `unit_rvalue` method and rename it to `unit_operand`.

Comment (lcnr) [reply src/librustc_middle/mir/mod.rs:2008]: > I'd appreciate avoiding new inherent methods on MIR data types that refer to the TyCtxt.

I didn't actually think about why we would want this.
Afaict the goal is for `librustc_mir` to not depend on `librustc_middle` @Centril?
Will impl it using a free method.

Comment (lcnr) [reply src/librustc_middle/mir/mod.rs:2008]: @eddyb we need this function in https://github.com/rust-lang/rust/blob/42abbd8878d3b67238f3611b0587c704ba94f39c/src/librustc_mir/transform/promote_consts.rs#L838
So we can't just place it in `librustc_mir_build` :cry:

Comment (Centril) [reply src/librustc_middle/mir/mod.rs:2008]: More precisely, the goal is to separate out the MIR data types to its own crate.

Comment (eddyb) [reply src/librustc_middle/mir/mod.rs:2008]: @lcnr The method would be used only where `unit_rvalue` is used today.
For the other two places, just copy-paste, it's tiny anyway.

Comment (lcnr) [reply src/librustc_middle/mir/mod.rs:2008]: This function is only used once :man_shrugging: just removed it entirely

File: src/librustc_mir_build/build/expr/as_rvalue.rs [hunk 2]
```diff
@@ -225,7 +225,11 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
             }
             ExprKind::Assign { .. } | ExprKind::AssignOp { .. } => {
                 block = unpack!(this.stmt_expr(block, expr, None));
-                block.and(this.unit_rvalue())
+                block.and(Rvalue::Use(Operand::Constant(box Constant {
+                    span: expr_span,
+                    user_ty: None,
+                    literal: ty::Const::zero_sized(self.hir.tcx(), self.hir.tcx().types.unit),
```
Comment (eddyb) [inline src/librustc_mir_build/build/expr/as_rvalue.rs:231]: You need to write `this` instead of `self`, note how the old code used `this.unit_rvalue()`.

**Comment by eddyb** [comment]
#70721 landed, you need to rebase and run `./x.py test --stage 1 src/test/mir-opt --bless`.

File: src/librustc_mir_build/build/expr/as_rvalue.rs (hunk 2 above)
Comment (lcnr) [reply src/librustc_mir_build/build/expr/as_rvalue.rs:231]: I just used copy paste for now, seems like I missed this :laughing:

**Comment by lcnr** [comment]
done

File: src/test/mir-opt/address-of/rustc.address_of_reborrow.SimplifyCfg-initial.after.mir [hunk 3]
```diff
@@ -298,7 +298,13 @@ fn address_of_reborrow() -> () {
...
         FakeRead(ForLet, _47);           // bb0[158]: scope 13 at $DIR/address-of.rs:36:9: 36:10
         AscribeUserType(_47, o, UserTypeProjection { base: UserType(29), projs: [] }); // bb0[159]: scope 13 at $DIR/address-of.rs:36:12: 36:22
-        _0 = ();                         // bb0[160]: scope 0 at $DIR/address-of.rs:3:26: 37:2
+        _0 = const ();                   // bb0[160]: scope 0 at $DIR/address-of.rs:3:26: 37:2
+                                         // ty::Const
+                                         // + ty: ()
+                                         // + val: Value(Scalar(<ZST>))
+                                         // mir::Constant
+                                         // + span: $DIR/address-of.rs:3:26: 37:2
+                                         // + literal: Const { ty: (), val: Value(Scalar(<ZST>)) }
```
Comment (eddyb) [inline src/test/mir-opt/address-of/rustc.address_of_reborrow.SimplifyCfg-initial.after.mir:307]: @rust-lang/wg-mir-opt Should we prioritize removing these verbose comments? I'm not sure they add much value.
(There's the `Span` but that only makes sense when it differs the statement `Span`, which it doesn't here)

**Comment by eddyb** [comment]
Blocking this on #70989, because we want to be able to tell on the PR itself that `32bit` wasn't updated.

File: src/librustc_mir_build/build/cfg.rs [hunk 4]
```diff
@@ -58,12 +59,17 @@ impl<'tcx> CFG<'tcx> {
...
     ) {
         self.push_assign(
             block,
             source_info,
             place,
-            Rvalue::Aggregate(box AggregateKind::Tuple, vec![]),
+            Rvalue::Use(Operand::Constant(box Constant {
+                span: source_info.span,
+                user_ty: None,
+                literal: ty::Const::zero_sized(tcx, tcx.types.unit),
```
Comment (eddyb) [inline src/librustc_mir_build/build/cfg.rs:71]: FWIW, I'm coincidentally adding `tcx.consts.unit` to #70970.

**Comment by eddyb** [comment]
@lcnr If you rebase, I think CI should now show the 32-bit mismatches.

**Comment by lcnr** [comment]
done

**Comment by eddyb** [comment]
@lcnr You manually fix it based on errors on CI (hard) or you run the `--bless` command with `--target i686-unknown-linux-gnu` (you may need "multilib" versions of your native toolchain e.g. GCC, if you get errors while compiling libstd).

**Comment by lcnr** [comment]
Ok, `--target i686-unknown-linux-gnu` was less of a hassle than I expected :+1: 

Ready to merge afaict.

**Review by eddyb — APPROVED** [review]

**Pull request merged**